erpnext/payroll/doctype/salary_slip/salary_slip.js

```diff
@@ -166,3 +200,89 @@ var total_work_hours = function(frm, dt, dn) {
 		refresh_many(['net_pay', 'rounded_total']);
 	});
 }
+
+var calculate_totals = function(doc) {
+	debugger;
+	var tbl1 = doc.earnings || [];
+	var tbl2 = doc.deductions || [];
+
+	var total_earn = 0; var total_ded = 0;
+	for(var i = 0; i < tbl1.length; i++){
+		total_earn += flt(tbl1[i].amount);
+	}
+	for(var j = 0; j < tbl2.length; j++){
+		total_ded += flt(tbl2[j].amount);
+	}
+	doc.gross_pay = total_earn;
+	doc.total_deduction = total_ded;
+	doc.net_pay = 0.0
+	if(doc.salary_slip_based_on_timesheet == 0){
+		doc.net_pay = flt(total_earn) - flt(total_ded) - flt(doc.total_loan_repayment);
+		doc.rounded_total = Math.round(doc.net_pay)
+	}
+	refresh_many(['gross_pay', 'total_deduction', 'net_pay', 'rounded_total']);
+}
+
+cur_frm.cscript.validate = function(doc, cdt, cdn) {
+	calculate_totals(doc);
+}
+
+
+frappe.ui.form.on('Salary Detail', {
+	amount: function(frm) {
+		calculate_totals(frm.doc);
+	},
+
+	earnings_remove: function(frm) {
+		calculate_totals(frm.doc);
+	},
+
+	deductions_remove: function(frm) {
+		calculate_totals(frm.doc);
+	},
+
+	salary_component: function(frm, cdt, cdn) {
```

Comment:
https://github.com/frappe/erpnext/wiki/Coding-Standards#all-business-logic-must-be-api-friendly-and-must-be-implemented-in-server-side-python